docs/enhancements/fep-001-api-fleet-templating.md

```diff
@@ -0,0 +1,303 @@
+```
+
+* Enable different folders within a config source based on labels:
+
+```yaml
+- name: microshift-manifests
+  configType: GitConfigProviderSpec
+  gitRef:
+    repository: basic-nginx-demo
+    targetRevision: 1.0.0
+    path: "/basic-nginx-demo/configuration/{{ index .device.metadata.labels `region` }}/{{  index .device.metadata.labels `site` }}"  
+```
+
+* Enable templating of inline configuration based on device details:
+```yaml
+ - name: motd-update
+   configType: InlineConfigProviderSpec
+   inline:
+     ignition:
+       version: 3.4.0
+     storage:
+       files:
+         - contents:
+             source: >-
+               data:,Device%20{{ .device.metadata.name }}%0AThis%20system%20is%20managed%20by%20flightctl.%0A
+           mode: 422
+           overwrite: true
+           path: "/etc/motd"
+```
+### Non-Goals
+
+## Proposal
+
+### Device API
+
+The device API does not change, in the device API we don't allow templating. Any
+device specifications from a fleet should already be rendered into the device
+
+Variable/template expansion is not supported at the device API level, as:
+  * We are operating on a single device, variables aren't really beneficial
+  * Updates to labels will not affect configuration unless the device is part of a fleet.
+
+### Fleet API
+
+The template.spec field of a fleet allows the use of go-template logic to render the device spec.
+```yaml
+kind: Fleet
+metadata:
+  name: autonomous-forklifts
+spec:
+  template:
+    spec:
+      os:
+        image: forklifts-{{ .device.metadata.label[factory] }}:latest
+      config:
+        - name: factory-wifi-access-credentials
+          configType: KubernetesSecretProviderSpec
+          k8sSecretRef:
+            cluster: …
+            namespace: factory-{{ .device.metadata.label[factory] }}
+            name: wifi-access-credentials
+            mountPath: /etc/network/…
+        - name: motd-update
+          configType: InlineConfigProviderSpec
+          inline:
+            ignition:
+              version: 3.4.0
+            storage:
+              files:
+                - contents:
+                    source: >-
+                        data:,{{ .device.name }}%0A
+                  mode: 422
+                  overwrite: true
+                  path: "/etc/acm-cluster-id"
+```
+
+The exposed values for use in templates is intentionally limited:
+
+* device.metadata.labels
+* device.metadata.name
+
+Making sure that the render of the device template is deterministic (i.e. not including timestamps or dates, etc..).
+
+#### Expected behavior
+
+When `fleet.spec.template` is updated, the fleet controller
+freezes the template, including non-expanded variables in the TemplateVersion. All existing devices
+are scanned to determine possible git/image/secret sources references by the `TemplateVersion` in the
+fleet. Those source references are captured in/alongside the `TemplateVersion` object. For non-versioned
```

Comment:
I know I pushed for this initially, but I don't think it's scalable.  If we have 10K devices in a fleet, and some secret has {{device.metadata.name}} in it, we'll have 10K entries in the TemplateVersion and validate 10K secrets before proceeding.  Obviously the example is a bit extreme, but you get the point.  Maybe we should validate the fleet template with best effort, ignoring any config items that have parameters.  Then we might fail to render devices and will see the error message on each relevant device.  Devices won't have downtime, however, because they will still see the latest successfully-rendered config.